modAionImpl/src/org/aion/zero/impl/sync/FastSyncManager.java

```diff
@@ -14,15 +37,56 @@
  */
 public final class FastSyncManager {
 
+    // TODO: ensure correct behavior when disabled
     private boolean enabled;
+    // TODO: ensure correct behavior when complete
     private final AtomicBoolean complete = new AtomicBoolean(false);
+    private final AtomicBoolean completeBlocks = new AtomicBoolean(false);
+
+    private final AionBlockchainImpl chain;
+    private final BlockHeaderValidator<A0BlockHeader> blockHeaderValidator;
+    private final P2pMgr p2pMgr;
+
+    // TODO: consider adding a FAST_SYNC log as well
+    private static final Logger log = AionLoggerFactory.getLogger(LogEnum.SYNC.name());
+
+    private AionBlock pivot = null;
+
+    Map<ByteArrayWrapper, Long> importedBlockHashes =
+            Collections.synchronizedMap(new LRUMap<>(4096));
+    Map<ByteArrayWrapper, ByteArrayWrapper> receivedBlockHashes =
+            Collections.synchronizedMap(new LRUMap<>(1000));
+
+    BlockingQueue<BlocksWrapper> downloadedBlocks = new LinkedBlockingQueue<>();
+    Map<ByteArrayWrapper, BlocksWrapper> receivedBlocks = new HashMap<>();
 
     private final Map<ByteArrayWrapper, byte[]> importedTrieNodes = new ConcurrentHashMap<>();
 
-    public FastSyncManager() {
-        this.enabled = false;
+    public FastSyncManager(
+            AionBlockchainImpl chain,
+            BlockHeaderValidator<A0BlockHeader> blockHeaderValidator,
+            final P2pMgr p2pMgr) {
+        this.enabled = true;
+        this.chain = chain;
+        this.blockHeaderValidator = blockHeaderValidator;
+        this.p2pMgr = p2pMgr;
     }
 
+    @VisibleForTesting
+    void setPivot(AionBlock pivot) {
+        Objects.requireNonNull(pivot);
+
+        this.pivot = pivot;
+    }
+
+    public AionBlock getPivot() {
+        return pivot;
+    }
+
+    // TODO: shutdown pool
+    ExecutorService executors =
+            Executors.newFixedThreadPool(Runtime.getRuntime().availableProcessors());
```

Comment:
Hmmm... isn't that overkill?

Reply:
Actually one of the most expensive CPU operations here is validating the PoW for the received blocks. This is why I aim at parallelizing this from the start.

Reply:
The number of threads used is still to be determined.